---
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
title: '`r params$title`'
output: 
  html_document:
    theme: flatly
  css: datatable_fix.css
params:
  GEMINI_data_frame:
    value: NA
  peddy_path_prefix:
    value: NA
  sample_peddy_id:
    value: NA
  sample_name:
    value: FILL_IN
  title:
    value: SampleReport
---

```{r, include = FALSE}
knitr::opts_chunk$set(echo = F,
                      warning = F,
                      message = F)
```

```{r, include = FALSE}
# if data given as character, assume it's a path to an Rdata file
if (is.character(params$GEMINI_data_frame)){
  gdf <- load(params$GEMINI_data_frame)
  GEMINI_data <- get(gdf)
  rm(gdf)
}
document_data <- See_GEM_formatter(GEMINI_data)
```


## `r params$sample_name` {.tabset}

### GEMINI Variation
```{r}

DT::datatable(document_data$GEMINI_data %>% mutate_if(is.numeric, funs(as.numeric(formatC(., format = "e", digits = 2)))), #width=1200, 
              escape = FALSE,
              rownames = F, 
              class='compact', 
              filter='bottom',
              extensions = c('FixedColumns','Buttons','FixedHeader'),
              options = list(columnDefs=list(list(targets=document_data$neg_core_index-1, visible = FALSE)),
                             pageLength = 15,
                             fixedHeader = TRUE,
                             lengthMenu = list(c(5, 15, 30, 60, 100, -1), list('5', '15', '30', '60', '100', 'All')),
                             dom = 'Bfrtip',
                             buttons = list(list(extend='pageLength'),
                                            list(extend='colvis', collectionLayout='four-column'),
                                            list(extend='colvisGroup', text='Core Fields', 
                                                 hide=paste(document_data$neg_core_index-1, collapse = ','), 
                                                 show = paste(document_data$core_index-1, collapse = ',')),
                                            list(extend='colvisGroup', text='Show All', show = ':hidden' )),
                             columnDefs=list(list(targets=c(1), class="dt-left")))) 

```

### peddy QC

```{r, include = FALSE}
# load peddy data
ped_check <- read_csv(paste0(params$peddy_path_prefix), "/ped_check.csv")
het_check <- read_csv(paste0(params$peddy_path_prefix), "/het_check.csv")
sex_check <- read_csv(paste0(params$peddy_path_prefix), "/sex_check.csv")
```

peddy Heterozygousity Check

mean/median_depth: mean/median depths for the sites tested.

het_ratio: proportion of sites that were heterozygous.

idr_baf: inter-decile range (90th percentile - 10th percentile) of b-allele frequency. We make a distribution of all sites of alts / (ref + alts) and then report the difference between the 90th and the 10th percentile. Large values indicated likely sample contamination.

peddy Pedigree Check

rel: the relatedness calculated from the genotypes.

ibs0: the number of sites at which the 2 samples shared no alleles (should approach 0 for parent-child pairs).
ibs2: the number of sites and which the 2 samples where both hom-ref, both het, or both hom-alt.

peddy Sex Check

ped_sex: sex from .ped file
predicted_sex: sex predicted from rate of hets on chrX.

### Info

This report is auto-generated with the [See GEM](https://github.com/davemcg/See_GEM) R package on `r strftime(Sys.time(), format = "%B %d, %Y")`

The analyst for this report populated the information in this dynamic documents with a variety of GEMINI queries. 

**FAQ**:

* What does it mean if a variant is marked in red (CURRENTLY NOT IMPLEMENTED)?
* This means that the variant has an impact severity of [HIGH](http://gemini.readthedocs.io/en/latest/content/database_schema.html#details-of-the-impact-and-impact-severity-columns) **or** is marked as pathogenic in ClinVar **and** has a population allele frequency less than 10%. 

**WARNINGS/LIMITATIONS** 


